lib/daru/category.rb

```diff
@@ -0,0 +1,235 @@
+module Daru
+  module Category
+    attr_accessor :coding_scheme, :base_category
+    attr_reader :index
+
+    def initialize_category data, opts={}
+      @type = :category
+
+      # Create a hash to map each category to positional indexes
+      categories = data.each_with_index.group_by(&:first)
+      @cat_hash = categories.map { |cat, group| [cat, group.map(&:last)] }.to_h
+
+      # Map each category to a unique integer for effective storage in @array
+      map_cat_int = categories.keys.each_with_index.to_h
+
+      # Inverse mapping of map_cat_int
+      @map_int_cat = map_cat_int.invert
+
+      # To link every instance to its category,
+      # it stores integer for every instance representing its category
+      @array = map_cat_int.values_at(*data)
+
+      # Specify if the categories are ordered or not.
+      # By default its unordered
+      @ordered = opts[:ordered] || false
+
+      # The coding scheme to code with. Default is dummy coding.
+      @coding_scheme = :dummy
+
+      # Base category which won't be present in the coding
+      @base_category = @cat_hash.keys.first
+
+      # Stores the name of the vector
+      @name = opts[:name]
+
+      # Index of the vector
+      @index = preprocess_index(opts[:index])
+    end
+
+    def each
+      return enum_for(:each) unless block_given?
+      @array.each { |pos| yield @map_int_cat[pos] }
+      self
+    end
+
+    def to_a
+      each.to_a
+    end
+
+    def size
+      @array.size
+    end
+
+    def ordered?
+      @ordered
+    end
+
+    def ordered= bool
+      @ordered = bool
+    end
+
+    def categories
+      @cat_hash.keys
+    end
+
+    def categories= new
+      # Change to SizeError
+      if new.size != categories.size
+        raise ShapeError, "New categories size (#{new.size}) does not match"\
+          "with old categories size (#{categories.size})"
+      end
+
+      @cat_hash = new.each_with_index.map do |new_cat, i|
+        @map_int_cat[i] = new_cat
+
+        old_cat = categories[i]
+        [new_cat, @cat_hash[old_cat]]
+      end.to_h
+    end
+
+    def order= new
+      if new.to_set != categories.to_set
+        raise ArgumentError, 'The context of new and old order must be the same.'
+      end
+
+      @cat_hash = new.map { |cat| [cat, @cat_hash[cat]] }.to_h
+    end
+
+    def min
```

Comment:
I'm thinking about extending the [Arel like query syntax](http://v0dro.github.io/blog/2015/08/03/finding-and-combining-data-in-daru/) which is already in Daru for that purpose.